src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -208,21 +211,61 @@ def dependencies(
         return (entry, tuple(entries[(i.group, i.artifact)] for i in entry.dependencies))
 
     @classmethod
-    def from_json_dict(cls, lockfile) -> CoursierResolvedLockfile:
+    def from_json_dicts(cls, json_lock_entries) -> CoursierResolvedLockfile:
         """Construct a CoursierResolvedLockfile from its JSON dictionary representation."""
 
-        return cls(entries=tuple(CoursierLockfileEntry.from_json_dict(dep) for dep in lockfile))
+        return cls(
+            entries=tuple(CoursierLockfileEntry.from_json_dict(dep) for dep in json_lock_entries)
+        )
 
-    def to_json(self) -> bytes:
-        """Export this CoursierResolvedLockfile to human-readable JSON.
+    @classmethod
+    def from_toml(cls, lockfile: str | bytes) -> CoursierResolvedLockfile:
+        """Constructs a CoursierResolvedLockfile from it's TOML + metadata comment representation.
 
-        This JSON is intended to be checked in to the user's repo as a hermetic snapshot of a
-        Coursier resolved JVM classpath.
+        The toml file should consist of an `[entries]` block, followed by several entries.
         """
-        return json.dumps([entry.to_json_dict() for entry in self.entries], indent=4).encode(
-            "utf-8"
+
+        lockfile_str: str
+        if isinstance(lockfile, str):
+            lockfile_str = lockfile
+        else:
+            lockfile_str = lockfile.decode("utf-8")
+
+        contents = toml.loads(lockfile_str)
+        entries = tuple(
+            CoursierLockfileEntry.from_json_dict(entry) for entry in (contents["entries"])
         )
 
+        return cls(
+            entries=entries,
+        )
+
+    @classmethod
+    def from_serialized(cls, lockfile: str | bytes) -> CoursierResolvedLockfile:
+        """Construct a CoursierResolvedLockfile from its serialized representation (either TOML with
+        attached metadata, or old-style JSON.)."""
+
+        try:
+            return cls.from_toml(lockfile)
+        except toml.TomlDecodeError:
+            deprecated.warn_or_error(
+                "2.11.0dev", "JSON-encoded JVM lockfile", "run `./pants generate-lockfiles"
```

Comment:
```suggestion
                "2.11.0.dev0", "JSON-encoded JVM lockfile", "run `./pants generate-lockfiles"
```
Versions are validated, but only if the code runs.